File: 2024 April/Strictly Increasing Array/main.py
# Link: https://www.geeksforgeeks.org/problems/convert-to-strictly-increasing-array3351/1


class Solution:
    def min_operations(self, nums):
        n = len(nums)
        LIS = [1] * n
        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j] and (i - j) <= (nums[i] - nums[j]):
                    LIS[i] = max(LIS[i], LIS[j] + 1)

        return n - max(LIS)


if __name__ == "__main__":
    obj = Solution()
    nums = [1, 2, 3, 6, 5, 4]
    print(obj.min_operations(nums))
    nums = [1, 1, 1, 1]
    print(obj.min_operations(nums))

File: 2024 April/Strictly Increasing Array/main.cpp
// Link: https://www.geeksforgeeks.org/problems/convert-to-strictly-increasing-array3351/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int min_operations(vector<int> &nums)
    {
        int n = nums.size(), ans = 1;
        vector<int> dp(n, 1);
        for (int i = 1; i < n; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (i - j <= nums[i] - nums[j])
                    dp[i] = max(dp[i], dp[j] + 1);
            }
            ans = max(ans, dp[i]);
        }
        return n - ans;
    }
};

int main()
{
    Solution obj;
    vector<int> nums = {1, 2, 6, 5, 4};
    cout << obj.min_operations(nums) << endl;
    nums = {1, 1, 1, 1};
    cout << obj.min_operations(nums) << endl;
    return 0;
}